File: src/main/java/io/github/mdsimmo/bomberman/game/Explosion.java
package io.github.mdsimmo.bomberman.game;

import io.github.mdsimmo.bomberman.Bomberman;
import io.github.mdsimmo.bomberman.events.*;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.Sound;
import org.bukkit.block.Block;
import org.bukkit.block.BlockState;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.Plugin;

import java.util.*;
import java.util.stream.Collectors;

public class Explosion implements Listener {

	public static class BlockPlan {
		public final Block block;
		public final BlockState prior, ignited, destroyed;

		public BlockPlan(Block block, BlockState prior, BlockState ignited, BlockState destroyed) {
			this.block = block;
			this.prior = prior;
			this.ignited = ignited;
			this.destroyed = destroyed;
		}
	}

	private static final Plugin plugin = Bomberman.instance;

	public static boolean spawnExplosion(Game game, Location center, Player cause, int strength) {

		// Find where the explosion should expand to
		Set<Block> firePlanned = planFire(center, game, strength);
		Set<BlockPlan> plannedTypes = firePlanned.stream()
				.map(b -> {
					BlockState prior = b.getState();
					BlockState ignited = b.getState();
					ignited.setType(Material.FIRE);
					BlockState converted = b.getState();
					converted.setType(Material.AIR);
					return new BlockPlan(b, prior, ignited, converted);
				})
				.collect(Collectors.toSet());

		// Let others know
		BmExplosionEvent event = new BmExplosionEvent(game, cause, plannedTypes);
		Bukkit.getPluginManager().callEvent(event);
		if (event.isCancelled())
			return false;

		// Save the current state of all exploding blocks (for drops generation latter)
		var igniting = event.getIgniting();
		igniting.forEach(b -> b.ignited.update(true));
		Objects.requireNonNull(center.getWorld())
				.playSound(center, Sound.ENTITY_GENERIC_EXPLODE, 1,
				(float)Math.random() + 0.5f );

		// Add an explosion obj to handle cleanup/watching for kills
		Explosion explosion = new Explosion(game, igniting, cause);
		Bukkit.getPluginManager().registerEvents(explosion, plugin);
		return true;
	}

	public static boolean isTouching(Player player, Set<Block> blocks) {
		return blocks.stream().anyMatch(b -> {
			double margin = 0.295; // magical value that seems to be how far fire burns players
			Location l = player.getLocation();
			Location min = b.getLocation().add( 0, -1, 0 );
			Location max = b.getLocation().add( 1, 2, 1 );
			return 	   l.getX() >= min.getX() - margin
					&& l.getX() <= max.getX() + margin
					&& l.getY() >= min.getY() - margin
					&& l.getY() <= max.getY() + margin
					&& l.getZ() >= min.getZ() - margin
					&& l.getZ() <= max.getZ() + margin;
		});
	}

	/**
	 * creates fire in the '+' pattern
	 */
	private static Set<Block> planFire(Location center, Game game, int strength) {
		Set<Block> blocks = new HashSet<>();
		// arms
		blocks.addAll(planFire(center, game, strength, 0, 1));
		blocks.addAll(planFire(center, game, strength, 0, -1));
		blocks.addAll(planFire(center, game, strength, 1, 0));
		blocks.addAll(planFire(center, game, strength, -1, 0));

		// center column
		for (int i = -1; i <= 1; i++) {
			planFire(center, game, 0, i, 0, blocks);
		}
		return blocks;
	}

	/**
	 * creates a line of fire in the given x, z direction;
	 *
	 * @param xstep the unit to step in the x direction
	 * @param zstep the unit to step in the z direction
	 */
	private static Set<Block> planFire(Location center, Game game, int strength, int xstep, int zstep) {
		Set<Block> blocks = new HashSet<>();
		for (int i = 1; i <= strength; i++) {
			planFire(center, game, i * xstep, 1, i * zstep, blocks);
			planFire(center, game, i * xstep, -1, i * zstep, blocks);
			if (planFire(center, game, i * xstep, 0, i * zstep, blocks))
				return blocks;
		}
		return blocks;
	}

	/**
	 * creates fire at the given location if it can.
	 *
	 * @return true if the fire-ball should stop
	 */
	private static boolean planFire(Location center, Game game, int x, int y, int z, Set<Block> blocks) {
		Location l = center.clone().add(z, y, x);
		Block b = l.getBlock();

		// Pass through air/weak things
		if (b.isPassable()) {
			blocks.add(b);
			return false;
		}

		// If hit dirt, blow up one block and then stop
		if (game.getSettings().getDestructable().contains(b.getType()))
			blocks.add(b);
		return true;
	}

	private final Game game;
	private final Player cause;
	private final Set<BlockPlan> blocks;
	private final int taskId;

	private Explosion(Game game, Set<BlockPlan> blocks, Player cause) {
		this.game = game;
		this.cause = cause;
		this.blocks = blocks;

		this.taskId = Bukkit.getScheduler().scheduleSyncDelayedTask(plugin, () -> {
			// Replace fire with air.
			// Note: Check if fire first because a player may have placed a block onto the space.
			// We cannot just use state.update(false) because the state was captured taken BEFORE the block was
			// converted into fire, so that would always do nothing
			blocks.forEach(b -> {
				if (b.ignited.getType() == b.block.getType()) {
					b.destroyed.update(true);
				}
			});

			// Give player back their TNT
			// Check for tag in case they have left the game already
			if (cause.getScoreboardTags().contains("bm_player"))
				cause.getInventory().addItem(new ItemStack(game.getSettings().getBombItem(), 1));

			// Drop loot
			var dropsPlaned = planDrops();
			var lootEvent = new BmDropLootEvent(game, cause, blocks, dropsPlaned);
			Bukkit.getPluginManager().callEvent(lootEvent);
			if (!lootEvent.isCancelled()) {
				lootEvent.getDrops().forEach((location, items) -> items.forEach(item -> {
					if (item.getAmount() > 0) {
						Objects.requireNonNull(location.getWorld())
								.dropItem(location.clone().add(0.5, 0.5, 0.5), item);
					}
				}));
			}

			// Delete ths obj from memory
			HandlerList.unregisterAll(this);
		}, 20);

	}

	private Map<Location, Set<ItemStack>> planDrops() {
		Map<Material, Map<ItemStack, Number>> loot = game.getSettings().getBlockLoot();
		return blocks.stream()
				.map(b -> new AbstractMap.SimpleEntry<>(
						b.block.getLocation(),
						lootSelect(loot.getOrDefault(b.prior.getType(), Collections.emptyMap()))))
				.collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
	}

	static <T> Set<T> lootSelect(Map<? extends T, ? extends Number> loot) {
		double sum = loot.values().stream()
				.map(Number::doubleValue)
				.reduce(0.0, Double::sum);
		for (var entry : loot.entrySet()) {
			var item = entry.getKey();
			var weight = entry.getValue().doubleValue();
			if (sum * Math.random() <= weight) {
				return new HashSet<>(Set.of(item));
			}
			sum -= weight;
		}

		if (sum == 0)
			return new HashSet<>();
		else
			throw new RuntimeException("Explosion.drop didn't select (should never happen)");
	}

	@EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
	public void onPlayerMove(BmPlayerMovedEvent e) {
		if (e.getGame() != game)
			return;
		if (isTouching(e.getPlayer(), blocks.stream().map(b -> b.block).collect(Collectors.toSet()))) {
			BmPlayerHitIntent.hit(e.getPlayer(), cause);
		}
	}

	@EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
	public void onAnotherExplosion(BmExplosionEvent e) {
		// Don't double remove blocks
		blocks.removeIf(thisBlock -> e.getIgniting().stream()
				.anyMatch(eventBlock -> eventBlock.block.equals(thisBlock.block)));
	}

	@EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
	public void onRunStopped(BmRunStoppedIntent e) {
		if (e.getGame() != game)
			return;
		Bukkit.getScheduler().cancelTask(taskId);
	}
}

File: core/src/main/java/io/github/mdsimmo/bomberman/game/Explosion.kt
package io.github.mdsimmo.bomberman.game

import io.github.mdsimmo.bomberman.Bomberman
import io.github.mdsimmo.bomberman.events.BmDropLootEvent
import io.github.mdsimmo.bomberman.events.BmExplosionEvent
import io.github.mdsimmo.bomberman.events.BmPlayerHitIntent.Companion.hit
import io.github.mdsimmo.bomberman.events.BmPlayerMovedEvent
import io.github.mdsimmo.bomberman.events.BmRunStoppedIntent
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.Sound
import org.bukkit.block.Block
import org.bukkit.block.BlockState
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.inventory.ItemStack
import java.util.function.Consumer
import java.util.stream.Collectors

class Explosion private constructor(
        private val game: Game,
        private val blocks: MutableSet<BlockPlan>,
        private val cause: Player)
    : Listener {

    data class BlockPlan(
            val block: Block,
            val prior: BlockState,
            val ignited: BlockState,
            val destroyed: BlockState
    )

    private val taskId: Int = Bukkit.getScheduler().scheduleSyncDelayedTask(Bomberman.instance,
            { cleanup() }, game.settings.fireTicks.toLong())
    private var noExplode = false

    private fun cleanup() {
        if (noExplode) {
            HandlerList.unregisterAll(this)
            return
        }

        // Replace fire with air.
        // Note: Check if fire first because a player may have placed a block onto the space.
        // We cannot just use state.update(false) because the state was captured taken BEFORE the block was
        // converted into fire, so that would always do nothing
        blocks.forEach(Consumer { b: BlockPlan ->
            if (b.ignited.type == b.block.type) {
                b.destroyed.update(true)
            }
        })

        // Give player back their TNT
        // Check for tag in case they have left the game already
        if (cause.scoreboardTags.contains("bm_player"))
            cause.inventory.addItem(ItemStack(game.settings.bombItem, 1))

        // Drop loot
        val dropsPlaned = planDrops()
        val lootEvent = BmDropLootEvent(game, cause, blocks, dropsPlaned)
        Bukkit.getPluginManager().callEvent(lootEvent)
        if (!lootEvent.isCancelled) {
            lootEvent.drops.forEach { (location: Location, items: Set<ItemStack>) ->
                items.forEach(Consumer { item: ItemStack ->
                    if (item.amount > 0) {
                        location.world?.dropItem(location.clone().add(0.5, 0.5, 0.5), item)
                    }
                })
            }
        }

        // Delete this obj from memory
        HandlerList.unregisterAll(this)
    }

    private fun planDrops(): Map<Location, Set<ItemStack>> {
        val loot = game.settings.blockLoot
        return blocks
                .map { b: BlockPlan ->
                    Pair(
                            b.block.location,
                            lootSelect(loot[b.prior.type] ?: emptyMap())
                    )
                }.toMap()
    }

    @EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
    fun onPlayerMove(e: BmPlayerMovedEvent) {
        if (e.game != game) return
        if (isTouching(e.player, blocks.stream().map { b: BlockPlan -> b.block }.collect(Collectors.toSet()))) {
            hit(e.player, cause)
        }
    }

    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    fun onAnotherExplosion(e: BmExplosionEvent) { // Don't double remove blocks
        blocks.removeIf { thisBlock: BlockPlan ->
            e.igniting
                    .any { it.block == thisBlock.block }
        }
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    fun onRunStopped(e: BmRunStoppedIntent) {
        if (e.game != game) return

        Bukkit.getScheduler().cancelTask(taskId)
        HandlerList.unregisterAll(this)
        noExplode = true // need noExplode because sometimes tasks don't cancel?!?!
    }

    companion object {

        @JvmStatic
        fun spawnExplosion(game: Game, center: Location, cause: Player, strength: Int): Boolean {
            // Find where the explosion should expand to
            val firePlanned = planFire(center, game, strength)
            val plannedTypes = firePlanned
                    .map { b: Block ->
                        val prior = b.state
                        val ignited = b.state
                        if (!game.settings.passKeep.contains(b.type))
                            ignited.type = game.settings.fireType
                        val converted = b.state
                        if (!game.settings.passKeep.contains(b.type) && !game.settings.passRevert.contains(b.type))
                            converted.type = Material.AIR
                        BlockPlan(b, prior, ignited, converted)
                    }
                    .toSet()

            // Let others know
            val event = BmExplosionEvent(game, cause, plannedTypes)
            Bukkit.getPluginManager().callEvent(event)
            if (event.isCancelled)
                return false

            // Save the current state of all exploding blocks (for drops generation latter)
            val igniting = event.igniting
            igniting.forEach {
                it.ignited.update(true)
            }
            center.world?.playSound(center, Sound.ENTITY_GENERIC_EXPLODE, 1f,Math.random().toFloat() + 0.5f)

            // Add an explosion obj to handle cleanup and watch for kills
            val explosion = Explosion(game, igniting, cause)
            Bukkit.getPluginManager().registerEvents(explosion, Bomberman.instance)
            return true
        }

        @JvmStatic
		fun isTouching(player: Player, blocks: Set<Block>): Boolean {
            return blocks.any { b: Block ->
                // TODO why not just listen for onPlayerDamage with FIRE_TICK type?
                val margin = 0.295 // magical value that seems to be how far fire burns players
                val l = player.location
                val min = b.location.add(0.0, -1.0, 0.0)
                val max = b.location.add(1.0, 2.0, 1.0)
                l.x >= min.x - margin && l.x <= max.x + margin
                        && l.y >= min.y - margin && l.y <= max.y + margin
                        && l.z >= min.z - margin && l.z <= max.z + margin
            }
        }

        /**
         * creates fire in the '+' pattern
         */
        private fun planFire(center: Location, game: Game, strength: Int): Set<Block> {
            val blocks = mutableSetOf<Block>()
            // arms
            blocks.addAll(planFire(center, game, strength, 0, 1))
            blocks.addAll(planFire(center, game, strength, 0, -1))
            blocks.addAll(planFire(center, game, strength, 1, 0))
            blocks.addAll(planFire(center, game, strength, -1, 0))
            // center column
            for (i in -1..1) {
                planFire(center, game, 0, i, 0, blocks)
            }
            return blocks
        }

        /**
         * creates a line of fire in the given x, z direction;
         *
         * @param xstep the unit to step in the x direction
         * @param zstep the unit to step in the z direction
         */
        private fun planFire(center: Location, game: Game, strength: Int, xstep: Int, zstep: Int): Set<Block> {
            val blocks = mutableSetOf<Block>()
            for (i in 1..strength) {
                planFire(center, game, i * xstep, 1, i * zstep, blocks)
                planFire(center, game, i * xstep, -1, i * zstep, blocks)
                if (planFire(center, game, i * xstep, 0, i * zstep, blocks)) return blocks
            }
            return blocks
        }

        /**
         * creates fire at the given location if it can.
         *
         * @return true if the fire-ball should stop
         */
        private fun planFire(center: Location, game: Game, x: Int, y: Int, z: Int, blocks: MutableSet<Block>): Boolean {
            val l = center.clone().add(z.toDouble(), y.toDouble(), x.toDouble())
            val b = l.block

            // Pass through air/weak things
            if (isPassing(b, game.settings)) {
                blocks.add(b)
                return false
            }

            // If hit destructible, blow up one block and then stop
            if (game.settings.destructible.contains(b.type))
                blocks.add(b)
            return true
        }

        private fun isPassing(b: Block, settings: GameSettings): Boolean {
            val t = b.type
            return t == Material.AIR || t == settings.fireType
                    || (b.isPassable && !(settings.indestructible.contains(t) || settings.destructible.contains(t)))
                    || settings.passDestroy.contains(t)
                    || settings.passKeep.contains(t)
                    || settings.passRevert.contains(t)
        }

        @JvmStatic
		fun <T> lootSelect(loot: Map<out T, Number>): Set<T> {
            // Sum the weights
            var sum = loot.values
                    .sumByDouble(Number::toDouble)

            // Select the item based off weights
            for ((item, value) in loot) {
                val weight = value.toDouble()
                if (sum * Math.random() <= weight) {
                    return setOf(item)
                }
                sum -= weight
            }

            // Handle possible empty set
            return if (sum == 0.0)
                emptySet()
            else
                throw RuntimeException("Explosion.drop didn't select (should never happen)")
        }
    }
}